feat: Added email verification system

Nil2000 opened the pull request:
# ✨ Codu Pull Request 💻

Fixes #1049 

## Pull Request details

- Implemented email verification part if the user want's to change email.

## Any Breaking changes

- None

## Associated Screenshots

![image](https://github.com/user-attachments/assets/20aad2f7-5e95-4fe0-9441-2bed15c85bdd)
![image](https://github.com/user-attachments/assets/7923d1c7-d998-421d-a639-0ddf4743ff7e)
![image](https://github.com/user-attachments/assets/d59678ac-a3d4-434f-8ef1-a5c170993b54)
![image](https://github.com/user-attachments/assets/f91df7d6-1609-4623-ad98-868a057e81d6)

Nil2000 committed:
feat: Added email verification system

Nil2000 committed:
Changed title for email verify

Nil2000 commented:
@NiallJoeMaher @John-Paul-Larkin Let me know what you think about this.

NiallJoeMaher commented on app/api/update_email/route.ts:
```diff
@@ -0,0 +1,25 @@
+import { NextRequest, NextResponse } from "next/server";
+
+export async function POST(req: NextRequest, res: NextResponse) {
+  const { userId, newEmail } = await req.json();
```
This could let someone change someone elses email. You would need to grab the user/userId from the current users session. Also for mutating requests like this we use trpc. You could add this here: https://github.com/codu-code/codu/blob/develop/server/api/router/profile.ts

And you'll see the pattern we have for cleaning data with schemas and getting the user from `ctx.session.user`

Nil2000 replied:
Yes followed this way now.

NiallJoeMaher commented on app/api/update_email/route.ts:
```diff
@@ -0,0 +1,25 @@
+} from "@/utils/emailToken";
+import { NextRequest, NextResponse } from "next/server";
+
+export async function POST(req: NextRequest, res: NextResponse) {
+  const { userId, newEmail } = await req.json();
+
+  if (!newEmail || !userId)
+    return NextResponse.json({ message: "Invalid request" }, { status: 400 });
+
+  const token = generateEmailToken();
+  const expiresAt = new Date(Date.now() + 1000 * 60 * 60); //1 hour
+
+  await storeTokenInDb(userId, token, expiresAt, newEmail);
+
+  await sendVerificationEmail(newEmail, token);
+
+  return NextResponse.json(
+    { message: "Verification email sent" },
+    { status: 200 },
+  );
+}
```
Agreed! A try catch is useful here.

NiallJoeMaher commented on app/api/verify_email/route.ts:
```diff
@@ -0,0 +1,33 @@
+import {
```
On the verify email route we could probably create a GET call which takes the token ID and redirects the user to a success or an error page depending on what they do (that's the normal pattern)

So have a route that might be `/verify-email?token=eyJhbGciOiJIUzI1NiIsInR`

Nil2000 replied:
Done ✅

NiallJoeMaher commented on app/api/verify_email/route.ts:
```diff
@@ -0,0 +1,33 @@
+} from "@/utils/emailToken";
+import { NextRequest, NextResponse } from "next/server";
+
+export async function POST(req: NextRequest, res: NextResponse) {
+  const token = req.nextUrl.searchParams.get("token");
+  const { userId } = await req.json();
+
+  if (!token)
+    return NextResponse.json({ message: "Invalid request" }, { status: 400 });
+
+  const tokenFromDb = await getTokenFromDb(token, userId);
+
+  console.log("tokenFromDb", tokenFromDb);
+  if (!tokenFromDb.length)
+    return NextResponse.json({ message: "Invalid token" }, { status: 400 });
+
+  const { expiresAt, email } = tokenFromDb[0];
+  if (expiresAt < new Date())
+    return NextResponse.json({ message: "Token expired" }, { status: 400 });
+
+  await updateEmail(userId, email);
+
+  await deleteTokenFromDb(token);
+
+  return NextResponse.json(
+    { message: "Email successfully verified" },
+    { status: 200 },
+  );
+}
```
Good suggestions here!

NiallJoeMaher commented on app/verify_email/_client.tsx:
```diff
@@ -0,0 +1,106 @@
+import { useRouter, useSearchParams } from "next/navigation";
+import React, { useEffect, useState } from "react";
+
+function Content({ userId }: { userId: string }) {
```
Good comment again. Nitpick, but the url would look better if it was `verify-email`.

NiallJoeMaher commented on app/verify_email/_client.tsx:
```diff
@@ -0,0 +1,106 @@
+      setStatus("loading");
+
+      try {
+        const res = await fetch(`/api/verify_email?token=${token}`, {
+          method: "POST",
+          headers: {
+            "Content-Type": "application/json",
+          },
+          body: JSON.stringify({ userId }),
+        });
```
The user should be able to click a link to a GET route which will auto-magically validate the token (instead of having to post a request). This ensures that the user clicked their email always.

NiallJoeMaher commented on app/verify_email/page.tsx:
```diff
@@ -0,0 +1,30 @@
+
+export default async function Page() {
+  const session = await getServerAuthSession();
+  console.log("session", session);
```
+1

NiallJoeMaher commented on app/verify_email/page.tsx:
```diff
@@ -0,0 +1,30 @@
+    redirect("/not-found");
+  }
+
+  const existingUser = await db.query.user.findFirst({
+    columns: {
+      id: true,
+    },
+    where: (users, { eq }) => eq(users.id, session.user!.id),
+  });
+
```
+1

NiallJoeMaher commented on drizzle/0007_shocking_rachel_grey.sql:
```diff
@@ -0,0 +1,53 @@
+	"userId" text NOT NULL,
+	CONSTRAINT "EmailVerificationToken_token_unique" UNIQUE("token")
+);
+--> statement-breakpoint
```
This migration seems to drop a lot of indexes that we might need. Also the file name is different than the ones we usually have which are descriptive.

Nil2000 replied:
Yeah saw that. But it's also creating those later.

NiallJoeMaher reviewed (changes requested):
Wowza! Lots of code on this one so great work so far. I added a good few comments but nothing mental. You have the guts of everything here. Well done!

Nil2000 committed:
Solved few issues

Nil2000 committed:
Typo resolved

Nil2000 commented:
@NiallJoeMaher Can you check if any more changes require?

### Key changes
1. Changed Migration file name (Removing `snapshot` and migration file and then created new with proper name)
2. Removed `api/update_email/route.ts` and introduced trpc method for updating email.
3. Introduced few `trycatch` block
4. Changed `POST` to `GET` for `token` verification (verifying userId from `getServerAut()`)

NiallJoeMaher commented on drizzle/0007_added-emailtokenverification-schema.sql:
```diff
@@ -0,0 +1,53 @@
+	CONSTRAINT "EmailVerificationToken_token_unique" UNIQUE("token")
+);
+--> statement-breakpoint
+DROP INDEX IF EXISTS "BannedUsers_userId_key";--> statement-breakpoint
```
Not sure what is going on here because we seem to be deleting and re-adding indexes (I know this is autocreated so I need to test this locally).

Nil2000 replied:
Yeah even I was not sure why just creating a new schema allows to have a migration which deletes prev indexes and again adding them later

NiallJoeMaher commented on utils/emailToken.ts:
```diff
@@ -0,0 +1,126 @@
+            <a href="${verificationLink}" class="btn">Verify Email</a>
+            <p>Please note that this link is valid for 1 hour only. If it expires, you will need to request a new one.</p>
+            <p>If you did not create an account, please ignore this email.</p>
+            <p>Best regards, <br>The Codu Team</p>
```
Spellcheck - Codú (we need the accent)

Nil2000 replied:
Oh my bad 😅

Nil2000 replied:
Done ✅

NiallJoeMaher reviewed (changes requested):
A few errors to catch anywhere you are doing an async/await operation. But nearly there! Well done! Tonne of work.

Nil2000 committed:
Fixed Codú spell

Nil2000 committed:
Fixed some try catch

NiallJoeMaher commented on server/db/schema.ts:
```diff
@@ -113,6 +113,17 @@ export const verificationToken = pgTable(
+  token: text("token").unique().notNull(),
+  createdAt: timestamp("createdAt").defaultNow().notNull(),
+  expiresAt: timestamp("expiresAt").notNull(),
+  email: text("email").notNull(),
```
I think we will also want `.unique()` on the email, or else someone could take an existing account's email. We will want to make sure the email isn't in use as suggested by coderabbit AI too.

Nil2000 replied:
True

Nil2000 replied:
Done ✅

NiallJoeMaher commented on schema/profile.ts:
```diff
@@ -18,6 +18,7 @@ export const saveSettingsSchema = z.object({
   websiteUrl: z.string().url().optional().or(z.literal("")),
   emailNotifications: z.boolean().optional(),
   newsletter: z.boolean().optional(),
+  email: z.string().email().optional().or(z.literal("")),
```
Shouldn't the user always have an email or am I missing something? 🤔

Nil2000 replied:
Let me try without having this once. Should work without this. Also I believe this email not required for updating the form associated to the settings form.

Nil2000 replied:
Ok sorted out ✅

NiallJoeMaher commented on server/api/router/profile.ts:
```diff
@@ -126,4 +133,33 @@ export const profileRouter = createTRPCRouter({
+          });
+        }
+
+        const userId = ctx.session.user.id;
+
+        const token = generateEmailToken();
```
This is definitely needed!

NiallJoeMaher reviewed (changes requested):
A handful of more changes. This is a huge PR so hopefully the next one we can break into different steps to make it easier for you (and get you some Hacktoberfest points to make up for the pain of this one!)

Nil2000 committed:
Fixed some typos

Nil2000 commented:
### Key changes

- Introduced `checkIfEmailExists` to check email used by any user already. Will be checked everytime user tries to update with already used email.
- Now email schema would be `unique()`
- Error was not shown in `toast` previously (Updated `client` page of settings to support this)

Nil2000 committed:
Fixed some typos

Nil2000 committed:
Fixed some typos

NiallJoeMaher commented:
A few conflicts, if you could delete the generated migration file and the meta files that were generated, pull the changes and re-run with `npx drizzle-kit generate --name=email-verification`. It should fix any issues that are holding things up with the migration conflicts and the weirdness that was happening.

Nil2000 commented:
> A few conflicts, if you could delete the generated migration file and the meta files that were generated, pull the changes and re-run with `npx drizzle-kit generate --name=email-verification`. It should fix any issues that are holding things up with the migration conflicts and the weirdness that was happening.

Yeah sure. Let me do that once.

Edit : Done ✅

Nil2000 committed:
Merged recent changes from main repo

NiallJoeMaher commented:
Your `package.lock` is different than the main repo, so it should probably be undone, and make sure you are installing with the correct node.js version. And just in case it fails the check after that make sure you run `npm run prettier:fix` to ensure you don't get stopped at the next hurdle. Once the checks are passing I'll approve and merge .

Nil2000 committed:
Ran prettier:fix and fixed package-lock.json

Nil2000 committed:
Fixed package-lock.json

NiallJoeMaher commented on package-lock.json:
```diff
@@ -19126,4 +19126,4 @@
     }
   }
-}
+}
```
This could be a breaking change (I think it's one too many) would you unstage this file (I think Praveen on Discord left the command).

NiallJoeMaher reviewed (approved):
LGTM! Great work! 🌮 🦖
